yktrace/src/hwt/mapper.rs

```diff
@@ -134,59 +130,66 @@ impl<'a> HWTMapper<'a> {
         ret
     }
 
-    /// Obtain the next mapped block.
+    /// Map the *machine* blocks of the specified trace into LLVM IR blocks.
     ///
-    /// `last_yielded` must be a reference to the last successfully mapped block returned by the
-    /// iterator (or `None` if this is the first time you are calling `next`).
-    pub fn next(
+    /// The returned trace is guaranteed to have the following properties:
+    ///
+    ///  - The returned trace will never start with an unmappable code block. It is a given that
+    ///    every raw trace will start by returning out of the foreign code that turned on trace
+    ///    collection. We omit this prefix.
+    ///
+    ///  - To save memory, in the returned trace consecutive unmappable block entries will have
+    ///    been collapsed into one. An unknown block entry therefore means "one or more unmappable
+    ///    blocks".
+    pub fn map_trace(
         &mut self,
-        last_yielded: Option<&IRBlock>,
-    ) -> Option<Result<IRBlock, HWTracerError>> {
-        while self.buffered_irblocks.is_empty() {
-            let block = self.trace_iter.next();
-            if block.is_none() {
-                return None;
-            }
-            let block = block.unwrap();
-            if let Err(e) = block {
-                return Some(Err(e));
-            }
-            let irblocks = self.map_block(block.as_ref().unwrap());
+        mut trace_iter: &'a mut dyn Iterator<Item = Result<Block, HWTracerError>>,
+    ) -> Result<Vec<IRBlock>, HWTracerError> {
+        let mut ret: Vec<IRBlock> = Vec::new();
+
+        for block in &mut trace_iter {
+            let irblocks = self.map_block(&block?);
             if irblocks.is_empty() {
                 // The block is unmappable. Insert a IRBlock that indicates this, but only if the
-                // trace isn't empty. We also take care to collapse repeated unmappable blocks into
-                // a single unmappable IRBlock.
-                if let Some(blk) = last_yielded {
-                    if !blk.is_unmappable() {
-                        self.buffered_irblocks.push(IRBlock::unmappable());
-                    }
+                // trace isn't empty (we never report the leading unmappable code in a trace). We
+                // also take care to collapse consecutive unmappable blocks into one.
+                if let Some(last) = ret.last_mut() {
+                    if !last.is_unmappable() {
+                        ret.push(IRBlock::unmappable());
+                    } // otherwise don't push, thus collapsing repeated unmappable blocks.
```

Comment:
I tend to find this sort of comment easier to read with an `else` so:

```rust
if !last.is_unmappable() {
  ret.push(IRBlock::unmappable());
} else {
  // otherwise don't push, thus collapsing repeated unmappable blocks.
}
```

I don't feel strongly about it though.

Reply:
Fixed in e3ac9f932e6cb56003498490264cf643b0683310